### PR TITLE
Notify about recent pushes if a branch has no PR yet

### LINKED ISSUE #311: PR feature merge hints
After pushing a new branch to forked repository, on original repository or forked repository, It should have hints to create a new pull request.

### LINKED ISSUE #13196: Prompt users to create a new PR after pushing to a new branch
As a user, I would like to see a prompt on the home page or repository page to create a new PR if I recently pushed to a new branch for which no PR exists yet so that I save clicking on the "new PR" button on the repo page.

This could be done purely through the db without any git calls since we already have all the data in the `action` table.

I think we can always assume the user has the rights to create a PR if they have the right to push to a new branch in that repo, so no need to check that.
We would just need another check if the repo pushed to is a fork to check with the head repo if a PR exists there and not in the fork repo.

### LINKED ISSUE #23743: Message new Branch(es) are ready to be merged
### Feature Description

As a developer, I want the ability to create pull requests fast. Therefore, I create a branch and switch then to the UI to click the button create a pull request. To improve my workflow, I would appreciate it if there was a banner to quickly create Pull Requests, like there is on GitHub. 

**Notes:**
- these messages should only appear for really newly created Branches, e.x. 1 hour
- It is available on the main page of the code repository
- Example: Example Image

### Screenshots

_No response_

### PR DESCRIPTION
**Pull request by kolaente**
This PR adds a nice little message on the dashboard if the current user recently pushed to a branch which has no PR open yet, with a link to do so. If the user recently pushed to a fork, it will show a message as well, but the link will go to create a PR with the upstream repository instead.

This PR explicitly does not show a link to existing PRs after pushes to a branch to keep it simple.

### Screenshots

image

image

image

image

image

### TODO

* [x] Make sure it is working for both pushes to a new branch on the same repo and on a fork
* [x] Actually show the message on the dashboard
* [x] Hide pushes to the default branch since it does not make sense to create PRs for that _unless_ the push was made to a forked repo
* [x] Include the time when the push happened in the message
* [x] Add a screenshot here

Resolves #311, resolves #13196, resolves #23743

### CODE REVIEW

**Comment by CirnoT** [comment]
I'm worried it might be too cluttered when more than few message are displayed (especially with rather long time limit). How about moving it to repo page similar to how GH does it?

**Comment by lunny** [comment]
How about display it in the repository home page but not user's dashboard page. Once the user push to serval repositories, which one should we display?

**Comment by kolaente** [comment]
> I'm worried it might be too cluttered when more than few message are displayed (especially with rather long time limit). How about moving it to repo page similar to how GH does it?

@CirnoT I think this could be addressed with either a shorter time limit or limiting the max messages to 2 or 4 (or both)

I think github shows it on both the repo page and the dashboard. IIRC gitlab does that as well.

**Comment by lunny** [comment]
Fix #23743

**Comment by kolaente** [comment]
Would there be interest in picking this PR up again?

**Comment by lunny** [comment]
Maybe it should be on top of the repository home page but not dashboard?

**Comment by delvh** [comment]
The interest is definitely there, but I wonder if it wouldn't make more sense to create a new branch, one to resolve all conflicts that happened in the mean time, and two to give it the attention boost of a new PR?

File: options/locale/locale_en-US.ini
Comment (delvh) [inline options/locale/locale_en-US.ini]: According to @lunny's suggestion above, this message should then be renamed.

File: models/action.go
Comment (delvh) [inline models/action.go]: Hmm…
What's the limit on GitHub?
24 hours as well?

File: templates/user/recently_pushed_branches.tmpl
Comment (delvh) [inline templates/user/recently_pushed_branches.tmpl]: Should we extract the comparison perhaps into a separate helper function?
I can imagine that we need it at many places…

**Comment by kolaente** [comment]
@delvh Probably. Should I do that or wants someone else pick it up?

File: models/action.go
Comment (delvh) [inline models/action.go]: I think that's a too large limit.
What about `3` instead?

**Review by delvh — COMMENTED** [review]
I think it makes sense if you publish the new PR as you're also the author of this one.

Below you find some more things I just noticed.

File: models/action.go
Comment (kolaente) [reply models/action.go]: I'm not sure. I think 24h is a good base-line which we could adjust after feedback from users.

File: templates/user/recently_pushed_branches.tmpl
Comment (kolaente) [reply templates/user/recently_pushed_branches.tmpl]: You mean the link building?

File: models/action.go
Comment (kolaente) [reply models/action.go]: Sure

File: templates/user/recently_pushed_branches.tmpl
Comment (delvh) [reply templates/user/recently_pushed_branches.tmpl]: Yes

File: models/action.go
Comment (kolaente) [reply models/action.go]: Done.

File: templates/user/recently_pushed_branches.tmpl
Comment (kolaente) [reply templates/user/recently_pushed_branches.tmpl]: Looks like there's a function `CompareLink` which does exactly this.

**Comment by kolaente** [comment]
This is now ready for review.

**Review by yardenshoham — CHANGES_REQUESTED** [review]
CI failure is related

**Comment by yardenshoham** [comment]
image
I wonder if we should somehow restrict it to always display as a single line?

**Comment by kolaente** [comment]
>  I wonder if we should somehow restrict it to always display as a single line?

I've now changed it to this:

image

**Comment by yardenshoham** [comment]
image
Not sure why but the bold parts barely stand out in the dark theme, it might just be me though 🤷‍♂️

File: models/recently_pushed.go
Comment (yardenshoham) [inline models/recently_pushed.go:1]: I would greatly appreciate it if you could include some comments to help me better comprehend the underlying logic. Given my limited experience with golang/xorm, a more detailed explanation would be immensely helpful before I can confidently give my approval

Comment (kolaente) [reply models/recently_pushed.go:1]: Added some comments, please check back

**Comment by kolaente** [comment]
Now the tests are failing in postgres with this error message:

>     testlogger.go:77: 2023/04/01 20:25:58 ...ers/web/user/home.go:133:Dashboard() [E] [64289356-9] GetRecentlyPushedBranches: pq: invalid input syntax for type bigint: "issue.closed_unix"

I don't know what could cause the problem here. Both types being compared are bigints. Theoretically this should not be a problem.

**Review by yardenshoham — COMMENTED** [review]
Looks very good to me. Is there a reason to use blue instead of the usual theme? Maybe then the bold/strong parts will look better in the dark theme than they currently do.

I will approve if the build passes

File: models/recently_pushed.go
Comment (lunny) [inline models/recently_pushed.go]: You have to use `builder.Expr("action.created_unix = issue.closed_unix")`.

Comment (kolaente) [reply models/recently_pushed.go]: Done

**Comment by kolaente** [comment]
@yardenshoham Not really. I thought it looks good that way. I think it doesn't stand out enough if I don't use the blue style:

image

`primary` doesn't seem to work great either. Do you have another suggestion?

**Comment by yardenshoham** [comment]
How about green instead of blue to match Gitea's theme? Try to use the dark theme in your setup to see it how I do

**Comment by yardenshoham** [comment]
Maybe @silverwind or @wxiaoguang would know which CSS setup will fit here

**Review by yardenshoham — CHANGES_REQUESTED** [review]
CI failure is related

**Comment by kolaente** [comment]
I think just plain `green` looks pretty awful:

image

**Comment by yardenshoham** [comment]
Yeah, I guess we can stick with blue

**Comment by kolaente** [comment]
Took another look and it seems like there's a class called `positive` which seems to work well:

image

image

**Review by yardenshoham — APPROVED** [review]

**Comment by kolaente** [comment]
Tests are passing now :tada:

**Comment by kolaente** [comment]
@delvh @lunny Can you check again?

**Comment by yardenshoham** [comment]
@kolaente don't forget to update the PR description with the latest info and screenshots

**Comment by CirnoT** [comment]
Still think this belongs on individual repo page.

**Comment by kolaente** [comment]
> Still think this belongs on individual repo page.

I mean we could probably do both (as GitHub does it) but I think that should be a follow-up PR.

File: models/recently_pushed.go
Comment (wxiaoguang) [inline models/recently_pushed.go]: If the prefix is stripped, it's not a "RefName" anymore. It's a "BranchName"

File: web_src/css/home.css
Comment (wxiaoguang) [inline web_src/css/home.css]: I guess there are some helper styles, like `gt-ac`, `gt-p-0` and `gt-m-0`, then we do not need to add more styles.

Comment (kolaente) [reply web_src/css/home.css]: I thought that as well, but wasn't able to find them. Is there an overview with all these helper styles somewhere?

Comment (delvh) [reply web_src/css/home.css]: `web_src/css/helpers.css`

Comment (wxiaoguang) [reply web_src/css/home.css]: https://github.com/go-gitea/gitea/blob/main/web_src/css/helpers.css

Comment (kolaente) [reply web_src/css/home.css]: Thanks! I've now changed it to use the helper classes instead.

File: models/recently_pushed.go
Comment (kolaente) [reply models/recently_pushed.go]: Good catch! Changed.

**Comment by silverwind** [comment]
Agree, should be on repo page, but I guess I won't block it if you want it on frontpage. Would not do both positions.

**Comment by kolaente** [comment]
> Agree, should be on repo page, but I guess I won't block it if you want it on frontpage. Would not do both positions.

In my workflow, I'm pushing to a new branch and then just opening the root domain of my Gitea instance. If the message offering me to create a new PR from my recent changes is right there on the dashboard, it saves me a few clicks.

> I do see the risk of I push to many branches where I don't want to open a PR, it'll litter the frontpage with these.

That's why we only show the last three.

File: models/recently_pushed.go
Comment (silverwind) [inline models/recently_pushed.go]: 24 hours seems a bit long. I have a feeling this is much shorter on GitHub. Anyone knows how long it is? I would say 2-4 hours max.

Comment (silverwind) [reply models/recently_pushed.go]: As per https://stackoverflow.com/a/23129795/808699 it seems to be 2 hours.

**Comment by silverwind** [comment]
> In my workflow, I'm pushing to a new branch and then just opening the root domain of my Gitea instance. If the message offering me to create a new PR from my recent changes is right there on the dashboard, it saves me a few clicks.

With GH, I usually have the target repo open and expect the popup on either repo root, issue list or pr list. I would display it in all those three places there. I barely use the "home" page on either gitea or github.

File: models/recently_pushed.go
Comment (kolaente) [reply models/recently_pushed.go]: I've made it two hours now.

If there are people who want to change this in the future, we may add a config variable to do this.

**Comment by kolaente** [comment]
I've added it to the repo as well:

image

image

image

File: templates/user/recently_pushed_branches.tmpl
Comment (silverwind) [inline templates/user/recently_pushed_branches.tmpl]:
```suggestion
	<div class="ui positive message gt-p-0 gt-df gt-ac gt-fw">
		<div class="gt-f1">
			{{$timeSince := TimeSince .Time $.locale}}
			{{$.locale.Tr "repo.pulls.recently_pushed_to_branches" .Repo.Owner.Name .Repo.LowerName .BranchName $timeSince | Safe}}
		</div>
		<a aria-role="button" class="ui compact positive button gt-m-0" href="{{CompareLink .BaseRepo .Repo .BranchName}}">
		  {{$.locale.Tr "repo.pulls.compare_changes"}}
	  </a>
	</div>
```

Untested, please try first. It uses flexbox to push the button to the right. Much better than grid.

Comment (kolaente) [reply templates/user/recently_pushed_branches.tmpl]: That makes sense. It almost works, I've made it work based on your suggestion.

**Comment by silverwind** [comment]
Can you left-align the text? It seems off in those screenshots. Will probably test myself later today.

**Comment by kolaente** [comment]
> Can you left-align the text? It seems off in those screenshots. Will probably test myself later today.

Sure.

**Review by silverwind — DISMISSED** [review]

**Review by yardenshoham — CHANGES_REQUESTED** [review]
CI failure is related

**Comment by lunny** [comment]
I created a branch on the base repository from the UI, but no hint display.

**Comment by kolaente** [comment]
@lunny  It worked for me: 

image

Did you only create the branch? (without changes)? Where did you expect the message?

**Comment by kolaente** [comment]
The ci failure seems unrelated, can someone restart the build?

**Comment by yardenshoham** [comment]
`git commit -m "Trigger build" --allow-empty`

**Comment by lunny** [comment]
The SQL logs looks like this

```log
2023/04/07 15:40:47 ...s/recently_pushed.go:60:GetRecentlyPushedBranches() [I] [SQL] SELECT action.ref_name, action.repo_id, action.created_unix FROM "action" LEFT JOIN "pull_request" ON pull_request.head_branch = replace(action.ref_name, 'refs/heads/', '') LEFT JOIN "issue" ON pull_request.issue_id = issue.id LEFT JOIN "repository" ON action.repo_id = repository.id WHERE action.op_type=$1 AND action.act_user_id=$2 AND ((repository.default_branch != replace(action.ref_name, 'refs/heads/', '')) OR repository.is_fork=$3) AND (pull_request.id IS NULL OR (pull_request.has_merged=$4 AND issue.is_closed=$5 AND (action.created_unix > issue.closed_unix))) AND action.created_unix>=$6 GROUP BY action.ref_name, action.repo_id, action.created_unix ORDER BY "action"."created_unix" DESC LIMIT 3 [commit_repo 3 true false true 1680846047] - 13.430291ms
2023/04/07 15:40:47 ...s/recently_pushed.go:82:GetRecentlyPushedBranches() [I] [SQL] SELECT "id", "owner_id", "owner_name", "lower_name", "name", "description", "website", "original_service_type", "original_url", "default_branch", "num_watches", "num_stars", "num_forks", "num_issues", "num_closed_issues", "num_pulls", "num_closed_pulls", "num_milestones", "num_closed_milestones", "num_projects", "num_closed_projects", "num_action_runs", "num_closed_action_runs", "is_private", "is_empty", "is_archived", "is_mirror", "status", "is_fork", "fork_id", "is_template", "template_id", "size", "is_fsck_enabled", "close_issues_via_commit_in_any_branch", "topics", "trust_model", "avatar", "created_unix", "updated_unix" FROM "repository" WHERE (0=1 OR repository.id IN (SELECT repository.fork_id FROM repository WHERE 0=1)) [] - 2.824583ms
2023/04/07 15:40:47 ...s/recently_pushed.go:98:GetRecentlyPushedBranches() [I] [SQL] SELECT * FROM "user" LEFT JOIN "repository" ON "repository".owner_id = "user".id WHERE (0=1 OR repository.id IN (SELECT repository.fork_id FROM repository WHERE 0=1)) [] - 1.515667ms
```

**Comment by lunny** [comment]
I think this should be reimplemented after #22743 is merged. It will become very easy to get recently pushed branches after that PR.

**Review by lunny — CHANGES_REQUESTED** [review]
This PR's implementation has a big performance problem that should be resolved.

**Comment by kolaente** [comment]
> This PR's implementation has a big performance problem that should be resolved.

What exactly do you mean? What should be changed?

> I think this should be reimplemented after https://github.com/go-gitea/gitea/pull/22743 is merged. It will become very easy to get recently pushed branches after that PR.

Does that block this PR?

**Comment by wxiaoguang** [comment]
I think the problem could be partially resolved by this approach:

1. Only do a simple select query in action table (no JOIN). Since we only need the recent 2 hour activity record, and action table has indices, the performance should be good enough.
2. Then, use other queries to gather more details.

@lunny what do you think about this?

**Comment by kolaente** [comment]
But isn't the performance of one query with joins pretty much always better than multiple queries? The database knows a lot more about the data in the db than we will from the outside so the optimizer will probably do a better job on this.

**Comment by delvh** [comment]
Well, except you basically have a join over four of the largest tables in the whole Gitea code.
While it might work on small instances of up to 10 people, this is not really scalable for something of the level of `gitea.com` or even `codeberg.org`, where many millions of entries are in these tables.

**Comment by CirnoT** [comment]
> But isn't the performance of one query with joins pretty much always better than multiple queries? The database knows a lot more about the data in the db than we will from the outside so the optimizer will probably do a better job on this.

No, not really. `JOIN` is performed before filtering (as also suggested by order in which query is written and the fact that you can filter on joined fields) so if you need to join few tables with millions of results just to fetch ~100 latest results from single table and correlate data with others, it would be much more efficient to get individual IDs via singular index than to have database do full merge of million of records into temporary table and then query that.

**Comment by wxiaoguang** [comment]
> But isn't the performance of one query with joins pretty much always better than multiple queries? The database knows a lot more about the data in the db than we will from the outside so the optimizer will probably do a better job on this.

Theoretically yes, actually, it depends.

Databases are just "guessing" the best index, they may make mistakes. That's why MySQL has key word "USE INDEX" or even "FORCE INDEX", I have used them many times. PostgreSQL may be somewhat better, developers still could use some tricks to make it use some particular indices.

And that's why GitLab dropped other database support, only supports one database -- it's difficult to fine-tune a large dataset on different databases.

**Comment by CirnoT** [comment]
>Databases are just "guessing" the best index, they may make mistakes. That's why MySQL has key word "USE INDEX" or even "FORCE INDEX", I have used them many times. PostgreSQL may be somewhat better, developers still could use some tricks to make it use some particular indices.

This is more of a issue with certain engines than others. For example engines such as InnoDB do not really keep stats on exact count of rows and instead estimate them - this might lead to wrong index cardinality and wrong decision being taken.

**Comment by kolaente** [comment]
It looks like the `pull_request` table is the problem here. Explaining the query indicates it does a full table scan on that table:

```
+------+-------------+--------------+--------+----------------------------------------+-----------------------+---------+-----------------------------+------+-------------------------------------------------+
| id   | select_type | table        | type   | possible_keys                          | key                   | key_len | ref                         | rows | Extra                                           |
+------+-------------+--------------+--------+----------------------------------------+-----------------------+---------+-----------------------------+------+-------------------------------------------------+
|    1 | SIMPLE      | repository   | ALL    | PRIMARY,IDX_repository_is_fork         | NULL                  | NULL    | NULL                        | 359  | Using temporary; Using filesort                 |
|    1 | SIMPLE      | action       | ref    | IDX_action_au_r_c_u_d,IDX_action_r_u_d | IDX_action_au_r_c_u_d | 18      | const,gitea.repository.id   | 104  | Using index condition; Using where              |
|    1 | SIMPLE      | pull_request | ALL    | NULL                                   | NULL                  | NULL    | NULL                        | 5978 | Using where; Using join buffer (flat, BNL join) |
|    1 | SIMPLE      | issue        | eq_ref | PRIMARY                                | PRIMARY               | 8       | gitea.pull_request.issue_id | 1    | Using where                                     |
+------+-------------+--------------+--------+----------------------------------------+-----------------------+---------+-----------------------------+------+-------------------------------------------------+
```

And while this is indeed fast on my local dev instance, the query takes 500ms on my own Gitea instance (using mysql), which has 6456 PRs created to date (So it's not a large instance but not exactly a small one either).

My first guess was this is because there's no index on the `head_branch` column of the PR table. So I added one:

```sql
create index pull_request_action_head_branch on pull_request (head_branch);
```

Running the query again, it only took 31ms. Explaining the query again, it looks like the db now uses the index to fetch the PR details:

```
+------+-------------+--------------+--------+----------------------------------------+---------------------------------+---------+-----------------------------+------+------------------------------------+
| id   | select_type | table        | type   | possible_keys                          | key                             | key_len | ref                         | rows | Extra                              |
+------+-------------+--------------+--------+----------------------------------------+---------------------------------+---------+-----------------------------+------+------------------------------------+
|    1 | SIMPLE      | repository   | ALL    | PRIMARY,IDX_repository_is_fork         | NULL                            | NULL    | NULL                        | 359  | Using temporary; Using filesort    |
|    1 | SIMPLE      | action       | ref    | IDX_action_au_r_c_u_d,IDX_action_r_u_d | IDX_action_au_r_c_u_d           | 18      | const,gitea.repository.id   | 104  | Using index condition; Using where |
|    1 | SIMPLE      | pull_request | ref    | pull_request_action_head_branch        | pull_request_action_head_branch | 1023    | func                        | 2    | Using where                        |
|    1 | SIMPLE      | issue        | eq_ref | PRIMARY                                | PRIMARY                         | 8       | gitea.pull_request.issue_id | 1    | Using where                        |
+------+-------------+--------------+--------+----------------------------------------+---------------------------------+---------+-----------------------------+------+------------------------------------+
```

I'd love to see that experiment with the data from gitea.com.

I propose to add an index on the `pull_request` table and call it a day.

**Comment by kolaente** [comment]
> you need to join few tables with millions of results just to fetch ~100 latest results from single table and correlate data with others, it would be much more efficient to get individual IDs via singular index than to have database do full merge of million of records into temporary table and then query that.

Except the query starts by fetching the actions from the last two hours. If that's millions of rows _for the currently logged in user_ the database is either really beafy already (because then you probably have 100k+ users and need that much power) or it's not and everything else is slow as well. I'd argue this is very unlikely and thus not a problem.

> fields that perform very costly actions such as REPLACE() - this function is executed per every row before filter is applied!

But iirc only once for each result from the `action` table? See above, that shouldn't be that many in the first place.

**Comment by CirnoT** [comment]
> I'd love to see that experiment with the data from gitea.com.

Check this plan: https://explain.dalibo.com/plan/7fb7f63h6e0ca080

Some stats:

```
gitea=# select count(*) from pull_request;
 count 
-------
  6020
(1 row)

gitea=# select count(*) from action;
 count  
--------
 274991
(1 row)
```

The query was adjusted to fetch actions from about last 7 days to estimate how it would behave on moderately popular instance

```
gitea=# select count(*) from action where action.created_unix>=1680362902;
 count 
-------
  1183
(1 row)
```

**Comment by CirnoT** [comment]
Here's possibly better plan where query was adjusted for a user that commits more often (bot): https://explain.dalibo.com/plan/cab6gc122759b18c

**Comment by silverwind** [comment]
> And that's why GitLab dropped other database support, only supports one database -- it's difficult to fine-tune a large dataset on different databases.

That is my impression as well. Single database allows many optimizations and features that just are not possible with multi-database. We might consider it if we can provide migrations, but I think that will be not so easy.

**Comment by CirnoT** [comment]
Also on unrelated note, that query returns for quite weird results:
```
gitea=# SELECT action.ref_name, action.repo_id, action.created_unix FROM "action" LEFT JOIN "pull_request" ON pull_request.head_branch = replace(action.ref_name, 'refs/heads/', '') LEFT JOIN "issue" ON pull_request.issue_id = issue.id LEFT JOIN "repository" ON action.repo_id = repository.id WHERE action.op_type=5 AND action.act_user_id=32 AND ((repository.default_branch != replace(action.ref_name, 'refs/heads/', '')) OR repository.is_fork=TRUE) AND (pull_request.id IS NULL OR (pull_request.has_merged=FALSE AND issue.is_closed=TRUE AND (action.created_unix > issue.closed_unix))) AND action.created_unix>=1680362902 GROUP BY action.ref_name, action.repo_id, action.created_unix ORDER BY "action"."created_unix" DESC;
               ref_name               | repo_id | created_unix 
--------------------------------------+---------+--------------
 refs/heads/renovate/phpstan-monorepo |       1 |   1680460273
 refs/heads/renovate/phpstan-monorepo |       1 |   1680460272
 refs/heads/renovate/phpstan-monorepo |      13 |   1680460072
 refs/heads/renovate/webpack-5.x      |       6 |   1680459707
 refs/heads/renovate/phpstan-monorepo |       1 |   1680456696
 refs/heads/renovate/phpstan-monorepo |      13 |   1680456540
 refs/heads/renovate/phpstan-monorepo |       9 |   1680456286
 refs/heads/renovate/phpstan-monorepo |       9 |   1680456285
```
These branches do not exist, Multiple PRs for them exists but they are either merged or closed (unmerged).

**Comment by lunny** [comment]
> I think the performance problem could be partially resolved by this approach:
> 
>     1. Only do a simple select query in action table (no JOIN). Since we only need the recent 2 hour activity record, and action table has indices, the performance should be good enough.
> 
>     2. Then, use other queries to gather more details.
> 
> 
> @lunny what do you think about this?

I think we could also do some optimization at the moment but those are temporary methods. As I said, once branches have been synced into databases, this PR could be implemented very easily.  Assume we have a new table `branch` with an index `repo_id` and `updated` timestamp. We can easily get recently updated branch names of this repository or all the branches in this instance. No complex SQL, no Joins.

**Comment by silverwind** [comment]
We could on branch push push all pushed branches into some a append-only structure based on username where values would automatically expire after the 2 hour time period (like redis TTL command).

Then, when rendering, the operation is just a lookup in that list.

The only thing I'm not sure about if auto-expiring keys can be done with SQL, or if that would require a cache like redis.

**Comment by yardenshoham** [comment]
I think GitHub keeps it out of the initial page render. Maybe it can be done in the background with an API call and render the suggestion only when ready?

**Comment by silverwind** [comment]
Yeah, it is definitely async on GitHub but causes ugly JS pop-in. If we can find a performant synchronous solution it would be ideal.

**Comment by kolaente** [comment]
Thanks everyone for your investigations and queries. My goal was to use existing data structures to avoid a higher overhead for such a simple feature. Well, looks like this is not as simple as I thought.

> I think we could also do some optimization at the moment but those are temporary methods. As I said, once branches have been synced into databases, this PR could be implemented very easily. Assume we have a new table branch with an index repo_id and updated timestamp. We can easily get recently updated branch names of this repository or all the branches in this instance. No complex SQL, no Joins.

Should this PR be blocked and reworked when the other one got merged?

**Comment by yardenshoham** [comment]
* Blocked by https://github.com/go-gitea/gitea/pull/22743

**Comment by techknowlogick** [comment]
@yardenshoham caught up in a PR cleanup sweep, thanks for re-opening :)

**Comment by wxiaoguang** [comment]
I think maybe this PR could be taken in 1.20 if no objection for "only querying recent N hours `actions` records by a single/simple SELECT" ?

* If the query range is limited by time and the query is simple, it doesn't seem to cause server side load problem?

**Comment by silverwind** [comment]
Yes, and ideally only show notification on the repo page. Not on the frontpage.

**Comment by lunny** [comment]
Since #22743 merged, I think this could be continued now.

**Comment by silverwind** [comment]
As already mentioned, I would like to see this changed to only show on the repo page as it would be too annoying on the frontpage.

**Comment by kolaente** [comment]
I don't really have time to implement this in the next few weeks, so if anyone wants to pick it up, feel free. Otherwise I'll take a stab at it at the end of August / early September.

**Comment by 6543** [comment]
#25715

**Pull request closed without merging**
